### PR TITLE
[Design] Use IR to Generate Couler Program

### PR DESCRIPTION
**Pull request by typhoonzero**
Related: https://github.com/sql-machine-learning/sqlflow/pull/1146

Design about how to implement `codegen_couler.go`

### CODE REVIEW

File: doc/design/codegen_couler_use_ir.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+In this design, we'll only have `codegen_couler.go` to generate programs to run a SQL program, which will be translated into multiple couler "steps". Each step, we can execute a SQL statement, run a training python program or run prediction, etc.
+
+## Generate Couler Program Using IR
+
+For example, to run the training steps, we intend to use couler API call: `couler.{xgboost,tensorflow,elasticdl}.train(model_def, data)` to generate a training step. The `train` function must take lots of parameters to run the training job as the SQLFlow statement describes, see [here](https://github.com/sql-machine-learning/sqlflow/blob/develop/python/sqlflow_submitter/tensorflow/train.py#L52) as an example.
+
+To implement the single `codegen_couler.go` to support generate code that can run either Tensorflow/XGBoost/ElasticDL/ALPS programs, we have below two choices:
+
+1. `couler.{tensorflow/xgboost/elasticdl}.train` have different arguments defined, so we can do:
```
Comment (Yancey0623) [inline doc/design/codegen_couler_use_ir.md:15]: Maybe we can follow #1146 , keep the second way.

File: doc/design/codegen_couler_use_ir.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    ```go
+        filler := Filler{
+            ModelType: ir.ModelType,
+            TrainConfigCode : generateTrainConfig(ir),
+        }
+        tfTrainTemplate = `couler.{{.ModelType}}.train(train_config={{.TrainConfigCode}})`
+        // Do template rendering here.
+    ```
+    **Note** that the `train_config` dictionary is not "engine specific", so that it could be used by any engine. A sample `train_config` should look like:
```
Comment (tonyyang-svail) [inline doc/design/codegen_couler_use_ir.md:40]: LGTM. In this design, SQLFlow is only responsible for the parsing, verifying, feature derivation, and serializing the result into the following Python struct.

Comment (tonyyang-svail) [reply doc/design/codegen_couler_use_ir.md:40]: And we need to think about who is responsible for generating the user documentation. In my opinion, who is translating `train.epoch` to `epoch` should be responsible for docgen, which is Couler in this design. Do you think it is reasonable?

File: doc/design/codegen_couler_use_ir.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            "Features": {"sepal_length": {"type": "numeric", "shape": [1], "field_meta":{"name":"sepal_length", "dtype": "float32", "delimiter": "", "is_sparse": False}}...},
+            "Label": {"class": {"type": "numeric", "shape": [1], "field_meta": ...}}
+        }
+    ```
+
+## The Design Choice
+
+We intend to use the solution **No.2** described above for these reasons.
+
+1. If a data scientist needs to add a new type of engine (SVM, SKLearn, Pytorch, etc.) other than Tensorflow/XGBoost/ElasticDL, he/she can use python only to define a `couler.{new_engine}.train` function without modifying the Go code in SQLFlow.
```
Comment (tonyyang-svail) [inline doc/design/codegen_couler_use_ir.md:58]: This is indeed a great advantage. A data scientist is usually more familiar with Python than Go.

**Review by tonyyang-svail — APPROVED** [review]
LGTM.

File: doc/design/codegen_couler_use_ir.md (hunk 2 above)
Comment (typhoonzero) [reply doc/design/codegen_couler_use_ir.md:40]: @tonyyang-svail can't agree more. By the way, if couler is generating the documentation, we should put couler repo under github.com/sql-machine-learning so that we can generate documentation under sqlflow.org.
